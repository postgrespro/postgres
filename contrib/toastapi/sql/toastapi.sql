CREATE EXTENSION toastapi;

CREATE FUNCTION dummy_toaster_handler(internal)
RETURNS internal
AS '$libdir/toastapi'
LANGUAGE C;

-- add_toaster()
SELECT add_toaster(NULL, NULL);
SELECT add_toaster(NULL, 'foo');
SELECT add_toaster('bar', NULL);
SELECT add_toaster('', '');
SELECT add_toaster('foo', '');
SELECT add_toaster('foo', 'bar');
SELECT add_toaster('dummy', 'foo');

SELECT add_toaster('dummy', 'dummy_toaster_handler') AS dummy_toaster_oid
\gset

SELECT :dummy_toaster_oid <> 0;
SELECT add_toaster('dummy', 'foo');
SELECT :dummy_toaster_oid = add_toaster('dummy', 'dummy_toaster_handler');

SELECT add_toaster('foo', 'dummy_toaster_handler') AS foo_toaster_oid
\gset

SELECT :foo_toaster_oid <> 0;
SELECT :foo_toaster_oid <> :dummy_toaster_oid;
SELECT :foo_toaster_oid = add_toaster('foo', 'dummy_toaster_handler');

-- drop_toaster()
SELECT drop_toaster(NULL);
SELECT drop_toaster('');
SELECT drop_toaster('bar');
SELECT drop_toaster('foo') = :foo_toaster_oid;
SELECT drop_toaster('foo');


CREATE TABLE tab (id int, jb jsonb, b_comp bytea, b_uncomp bytea STORAGE external);

-- set_toaster()
SELECT set_toaster('', '', '');
SELECT set_toaster('foo', 'bar', 'baz');
SELECT set_toaster('foo', 'tab', 'baz');
SELECT set_toaster('dummy', 'tab', 'bar');
SELECT set_toaster('dummy', 'tab', 'id');
SELECT set_toaster('dummy', 'tab', 'jb');
SELECT set_toaster('dummy', 'tab', 'b_uncomp');
SELECT set_toaster('dummy', 'tab', 'b_comp') = :dummy_toaster_oid;

-- get_toaster()
SELECT get_toaster('', '');
SELECT get_toaster('foo', 'bar');
SELECT get_toaster('tab', 'bar');
SELECT get_toaster('tab', 'id');
SELECT get_toaster('tab', 'jb');
SELECT get_toaster('tab', 'b_uncomp');
SELECT get_toaster('tab', 'b_comp') = :dummy_toaster_oid;

-- reset_toaster()
SELECT reset_toaster('', '');
SELECT reset_toaster('foo', 'bar');
SELECT reset_toaster('tab', 'bar');
SELECT reset_toaster('tab', 'id');
SELECT reset_toaster('tab', 'jb');
SELECT reset_toaster('tab', 'b_uncomp');
SELECT reset_toaster('tab', 'b_comp');
SELECT get_toaster('tab', 'b_comp');
SELECT reset_toaster('tab', 'b_comp');
SELECT get_toaster('tab', 'b_comp');

SELECT set_toaster('dummy', 'tab', 'b_comp') = :dummy_toaster_oid;
SELECT get_toaster('tab', 'b_comp') = :dummy_toaster_oid;
SELECT reset_toaster('tab', 'b_comp');
SELECT get_toaster('tab', 'b_comp');
SELECT set_toaster('dummy', 'tab', 'b_comp') = :dummy_toaster_oid;

-- drop_toaster()
SELECT drop_toaster(NULL);
SELECT drop_toaster('');
SELECT drop_toaster('foo');
SELECT drop_toaster('dummy');
SELECT drop_toaster('dummy');
SELECT reset_toaster('tab', 'b_comp');
SELECT drop_toaster('dummy') = :dummy_toaster_oid;
SELECT drop_toaster('dummy');

DROP EXTENSION toastapi;
