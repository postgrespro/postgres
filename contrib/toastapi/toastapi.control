# toastapi extension
comment = 'toastapi - Pluggable TOAST API'
default_version = '1.0'
module_pathname = '$libdir/toastapi'
schema = pgpro_toast
relocatable = false
