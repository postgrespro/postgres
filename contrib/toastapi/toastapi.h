/*-------------------------------------------------------------------------
 *
 * toastapi.h
 *		Pluggable TOAST API
 *
 * Portions Copyright (c) 2016-2021, PostgreSQL Global Development Group
 * Portions Copyright (c) 1990-1993, Regents of the University of California
 *
 * IDENTIFICATION
 *	  contrib/toastapi/toastapi.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef TOASTAPI_H
#define TOASTAPI_H

#include "postgres.h"
#include "varatt.h"
#include "fmgr.h"
#include "access/heaptoast.h"
#include "access/htup_details.h"
#include "commands/defrem.h"
#include "utils/builtins.h"
#include "utils/syscache.h"
#include "access/toast_compression.h"
#include "access/attoptions.h"
#include "access/xact.h"
#include "catalog/binary_upgrade.h"
#include "catalog/catalog.h"
#include "catalog/dependency.h"
#include "catalog/heap.h"
#include "catalog/index.h"
#include "catalog/namespace.h"
#include "catalog/pg_am.h"
#include "catalog/pg_namespace.h"
#include "catalog/pg_opclass.h"
#include "catalog/pg_type.h"
#include "catalog/toasting.h"
#include "miscadmin.h"
#include "nodes/makefuncs.h"
#include "access/relation.h"
#include "access/genam.h"

#define TOASTER_HANDLEROID 8888

#define PG_TOASTER_NAME "pg_toaster"
#define PG_TOASTREL_NAME "pg_toastrel"

#define REL_TOASTER_NAME "toasteroid"
#define REL_BASEREL_NAME "basereloid"

#define ATT_TOASTER_NAME "toasteroid"
#define ATT_HANDLER_NAME "toasthandler"
#define ATT_TOASTREL_NAME "toastreloid"
#define ATT_NTOASTERS_NAME "ntoasters"
/*
static Oid pg_toaster_idx_oid = InvalidOid;
static Oid pg_toastrel_idx_oid = InvalidOid;
*/

/*
 * Macro to fetch the possibly-unaligned contents of an EXTERNAL datum
 * into a local "struct varatt_external" toast pointer.  This should be
 * just a memcpy, but some versions of gcc seem to produce broken code
 * that assumes the datum contents are aligned.  Introducing an explicit
 * intermediate "varattrib_1b_e *" variable seems to fix it.
 */
#define VARATT_EXTERNAL_GET_POINTER(toast_pointer, attr) \
do { \
	varattrib_1b_e *attre = (varattrib_1b_e *) (attr); \
	Assert(VARATT_IS_EXTERNAL(attre)); \
	Assert(VARSIZE_EXTERNAL(attre) == sizeof(toast_pointer) + VARHDRSZ_EXTERNAL); \
	memcpy(&(toast_pointer), VARDATA_EXTERNAL(attre), sizeof(toast_pointer)); \
} while (0)

/* Size of an EXTERNAL datum that contains a standard TOAST pointer */
#define TOAST_POINTER_SIZE (VARHDRSZ_EXTERNAL + sizeof(varatt_external))

/* Size of an EXTERNAL datum that contains an indirection pointer */
#define INDIRECT_POINTER_SIZE (VARHDRSZ_EXTERNAL + sizeof(varatt_indirect))

#define VARATT_TOASTER_GET_POINTER(toast_pointer, attr) \
do { \
	varattrib_1b_e *attre = (varattrib_1b_e *) (attr); \
	Assert(VARATT_IS_TOASTER(attre)); \
	Assert(VARSIZE_TOASTER(attre) == sizeof(toast_pointer) + VARHDRSZ_EXTERNAL); \
	memcpy(&(toast_pointer), VARDATA_TOASTER(attre), sizeof(toast_pointer)); \
} while (0)

/* Size of an EXTERNAL datum that contains a custom TOAST pointer */
#define TOASTER_POINTER_SIZE (VARHDRSZ_EXTERNAL + sizeof(varatt_custom))

#define BYTEA_TOASTER_MAGIC    0xb17ea757
typedef struct ByteaToastRoutine
{
       int32           magic;
       Datum     (*append)(Datum val1, Datum val2);
} ByteaToastRoutine;

typedef struct ToastAttributesData
{
	Oid toasteroid;
	Oid toasthandleroid;
	Oid toastreloid;
	int attnum;
	int ntoasters;
	void *toaster;
} ToastAttributesData;

typedef ToastAttributesData *ToastAttributes;

/*
 * Callback function signatures --- see indexam.sgml for more info.
 */

/* Create toast storage */
typedef Datum (*toast_init)(Relation rel, Oid toasteroid, Datum reloptions, int attnum, LOCKMODE lockmode,
						   bool check, Oid OIDOldToast, ToastAttributes tattrs);

/* Toast function */
typedef Datum (*toast_function) (Relation toast_rel,
										   Oid toasterid,
										   Datum value,
										   Datum oldvalue,
											int attnum,
										   int max_inline_size,
										   int options,
											ToastAttributes tattrs);

/* Update toast function, optional */
typedef Datum (*update_toast_function) (Relation toast_rel,
												  Oid toasterid,
												  Datum newvalue,
												  Datum oldvalue,
												  int options,
												  int attnum,
												  ToastAttributes tattrs);

/* Copy toast function, optional */
typedef Datum (*copy_toast_function) (Relation toast_rel,
												Oid toasterid,
												Datum newvalue,
												int options,
												int attnum,
												ToastAttributes tattrs);

/* Detoast function */
typedef Datum (*detoast_function) (Datum toast_ptr,
											 int offset, int length, ToastAttributes tattrs);

/* Delete toast function */
typedef void (*del_toast_function) (Relation rel,Datum value, bool is_speculative, ToastAttributes tattrs);

/* Return virtual table of functions, optional */
typedef void * (*get_vtable_function) (Datum toast_ptr);

/* validate definition of a toaster Oid */
typedef bool (*toastervalidate_function) (Oid typeoid,
										  char storage, char compression,
										  Oid amoid, bool false_ok);

/*
 * API struct for Toaster.  Note this must be stored in a single palloc'd
 * chunk of memory.
 */

typedef struct TsrRoutine
{
	NodeTag		type;

	/* interface functions */
	toast_init init;
	toast_function toast;
	update_toast_function update_toast;
	copy_toast_function copy_toast;
	detoast_function detoast;
	del_toast_function deltoast;
	get_vtable_function get_vtable;
	toastervalidate_function toastervalidate;
} TsrRoutine;

#define T_TsrRoutine 999
#define makeTsrNode()		((TsrRoutine *) newNode(sizeof(TsrRoutine),T_TsrRoutine))

/* Functions in toastapi.c */
extern TsrRoutine *GetTsrRoutine(Oid tsrhandler);
extern TsrRoutine *GetTsrRoutineByOid(Oid tsroid, bool noerror);
extern TsrRoutine *SearchTsrCache(Oid tsroid);
extern bool	validateToaster(Oid toasteroid, Oid typeoid, char storage,
							char compression, Oid amoid, bool false_ok);

extern Datum default_toaster_handler(PG_FUNCTION_ARGS);
extern bool get_toast_params(Oid relid, int attnum, ToastAttributes tattrs); // int *ntoasters, Oid *toasteroid, Oid *toastrelid, Oid *handlerid);

static inline void init_tattrs(ToastAttributes tattrs)
{
	tattrs = palloc(sizeof(ToastAttributesData));
	tattrs->attnum = -1;
	tattrs->ntoasters = 0;
	tattrs->toaster = NULL;
	tattrs->toasteroid = InvalidOid;
	tattrs->toasthandleroid = InvalidOid;
	tattrs->toastreloid = InvalidOid;
}

#endif							/* TOASTAPI_H */