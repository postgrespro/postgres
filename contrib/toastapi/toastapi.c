/*-------------------------------------------------------------------------
 *
 * toastapi.c
 *		Pluggable TOAST API
 *
 * Portions Copyright (c) 2016-2021, PostgreSQL Global Development Group
 * Portions Copyright (c) 1990-1993, Regents of the University of California
 *
 * IDENTIFICATION
 *	  contrib/toastapi/toastapi.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"
#include "varatt.h"
#include "fmgr.h"
#include "toastapi.h"
#include "access/toast_helper.h"

#include "access/htup_details.h"
#include "commands/defrem.h"
#include "lib/pairingheap.h"
#include "utils/builtins.h"
#include "utils/memutils.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "access/heaptoast.h"
#include "access/reloptions.h"
#include "access/attoptions.h"
#include "catalog/catalog.h"
#include "catalog/indexing.h"
#include "miscadmin.h"
#include "utils/fmgroids.h"
#include "utils/rel.h"
#include "utils/snapmgr.h"
#include "catalog/pg_namespace.h"
#include "utils/guc.h"

#include "catalog/binary_upgrade.h"
#include "catalog/dependency.h"
#include "catalog/heap.h"
#include "catalog/index.h"
#include "catalog/namespace.h"
#include "catalog/pg_am.h"
#include "catalog/pg_class.h"
#include "catalog/pg_opclass.h"
#include "catalog/pg_type.h"
#include "catalog/toasting.h"
#include "nodes/makefuncs.h"
#include "storage/lock.h"

#include "catalog/pg_am_d.h"
#include "commands/vacuum.h"
#include "funcapi.h"
#include "storage/bufmgr.h"

#include "libpq/auth.h"
#include "utils/guc.h"
#include "utils/timestamp.h"

#include "utils/lsyscache.h"
#include "utils/regproc.h"

#include "access/detoast.h"
#include "access/toast_internals.h"
#include "access/toast_extended.h"
#include "access/toast_hook.h"
#include "utils/elog.h"
#include "pg_toaster.h"
#include "pg_toastrel.h"
#include "utils/varlena.h"
#include "varatt_custom.h"
#include "toastapi_internals.h"
#include "toastapi_sqlfuncs.h"

PG_MODULE_MAGIC;

static Toastapi_init_hook_type toastapi_init_hook = NULL;
static Toastapi_toast_hook_type toastapi_toast_hook = NULL;
static Toastapi_detoast_hook_type toastapi_detoast_hook = NULL;
static Toastapi_size_hook_type toastapi_size_hook = NULL;
static Toastapi_copy_hook_type toastapi_copy_hook = NULL;
static Toastapi_update_hook_type toastapi_update_hook = NULL;
static Toastapi_delete_hook_type toastapi_delete_hook = NULL;

static Datum toastapi_init (Oid reloid, Datum reloptions, int attnum, LOCKMODE lockmode,
						   bool check, Oid OIDOldToast)
{
   Datum result = (Datum) 0;
	FormData_pg_attribute *pg_attr;
	Datum d;
	TsrRoutine *toaster = NULL;
	Relation rel;
	int ntoasters = 0;
	ToastAttributes tattrs;

	rel = table_open(reloid, RowExclusiveLock);
	pg_attr = &rel->rd_att->attrs[attnum];

	d = attopts_get_toaster_opts(reloid, NameStr(pg_attr->attname), attnum, ATT_NTOASTERS_NAME);
	if(d == (Datum) 0)
		return (Datum) 0;
	else
		ntoasters = atoi(DatumGetCString(d));

	{
		int len = 0;
		char str[12];

		len = pg_ltoa(ntoasters, str);
		d = get_complex_att_opt(RelationGetRelid(rel), ATT_HANDLER_NAME, str, len, attnum);
	}

	toaster = GetTsrRoutine(atoi(DatumGetCString(d)));
	if(toaster == NULL)
	{
		elog(NOTICE, "No routine found");
		return (Datum) 0;
	}

	tattrs = palloc(sizeof(ToastAttributesData));
	tattrs->toasteroid = InvalidOid;
	tattrs->toastreloid = InvalidOid;

	tattrs->attnum = attnum;
	tattrs->ntoasters = ntoasters;
	tattrs->toasthandleroid = atoi(DatumGetCString(d));
	tattrs->toaster = toaster;

	result = toaster->init(rel,
									atoi(DatumGetCString(d)),
									reloptions,
									attnum,
									lockmode,
									check,
									OIDOldToast,
									tattrs);
	table_close(rel, RowExclusiveLock);
	pfree(tattrs);
   return result;
}

static Datum toastapi_toast (ToastTupleContext *ttc, int attribute, int maxDataLen,
						int options)
{
   Datum result = (Datum) 0;
	Datum	   *value = &ttc->ttc_values[attribute];
	Datum		old_value = *value;
	ToastAttrInfo *attr = &ttc->ttc_attr[attribute];
	Datum d;
	Relation rel;
	TsrRoutine *toaster = NULL;
	char *ntoasters_str;
	Oid tsrhandler = InvalidOid;
	ToastAttributes tattrs;

	result = *value;
	rel = table_open(RelationGetRelid(ttc->ttc_rel), RowExclusiveLock);

	d = attopts_get_toaster_opts(RelationGetRelid(ttc->ttc_rel), "", attribute+1, ATT_NTOASTERS_NAME);
	
	if(d == (Datum) 0)
	{
		result = toast_save_datum(ttc->ttc_rel, old_value, attr->tai_oldexternal,
			options);

		table_close(rel, RowExclusiveLock);
		return result;
	}

	ntoasters_str = DatumGetCString(d);

	d = get_complex_att_opt(RelationGetRelid(rel), ATT_HANDLER_NAME, ntoasters_str, strlen(ntoasters_str), attribute+1);

	if(d == (Datum) 0)
	{
		result = toast_save_datum(ttc->ttc_rel, old_value, attr->tai_oldexternal,
			options);

		table_close(rel, RowExclusiveLock);
		return result;
	}
	else
	{
		tsrhandler = atoi(DatumGetCString(d));
		if(OidIsValid(tsrhandler))
			toaster = GetTsrRoutine(tsrhandler);
		else
		{
			result = toast_save_datum(ttc->ttc_rel, old_value, attr->tai_oldexternal, options);
			table_close(rel, RowExclusiveLock);
			return result;
		}
	}

	d = get_complex_att_opt(RelationGetRelid(rel), ATT_TOASTREL_NAME, ntoasters_str, strlen(ntoasters_str), attribute+1);

	table_close(rel, RowExclusiveLock);

	if(d == (Datum) 0)
	{
		toastapi_init(RelationGetRelid(ttc->ttc_rel), (Datum) 0, attribute+1, RowExclusiveLock, false, InvalidOid);
		rel = table_open(RelationGetRelid(ttc->ttc_rel), RowExclusiveLock);
		d = get_complex_att_opt(RelationGetRelid(rel), ATT_TOASTREL_NAME, ntoasters_str, strlen(ntoasters_str), attribute+1);
		table_close(rel, RowExclusiveLock);
	}

	if(toaster != NULL)
	{
		tattrs = palloc(sizeof(ToastAttributesData));
		tattrs->toasteroid = InvalidOid;
		tattrs->toastreloid = InvalidOid;

		tattrs->attnum = attribute;
		tattrs->ntoasters = atoi(ntoasters_str);
		tattrs->toasthandleroid = tsrhandler;
		tattrs->toaster = toaster;
		if(d != (Datum) 0)
			tattrs->toastreloid = atoi(DatumGetCString(d));

		result = toaster->toast(ttc->ttc_rel,
										tsrhandler,
										old_value,
										old_value, // PointerGetDatum(attr->tai_oldexternal),
										attribute+1,
										maxDataLen, options, tattrs);
		pfree(tattrs);
	}

	return result;
}

static Size toastapi_size (uint8 tag, const void *ptr)
{
	return offsetof(varatt_custom, va_toasterdata) + VARATT_CUSTOM_GET_DATA_SIZE(ptr);
}
/*
static Size toastapi_size (uint8 tag, const void *ptr)
{
	return (tag) == VARTAG_CUSTOM ? offsetof(varatt_custom, va_toasterdata) + VARATT_CUSTOM_GET_DATA_SIZE(ptr) : sizeof(ptr);
}
*/
static Datum toastapi_detoast (Oid relid, Datum toast_ptr,
											 int offset, int length)
{
	struct varlena *value;
	Datum result = toast_ptr;

	value = (struct varlena *) DatumGetPointer(toast_ptr);
	if(VARATT_IS_EXTERNAL_ONDISK(value))
	{
		elog(NOTICE, "Detoast hook called for regular External TOAST pointer");
	}
	if(VARATT_IS_CUSTOM(value))
	{
		TsrRoutine *toaster = NULL;
		ToastAttributes tattrs;
		Oid	toasterid = VARATT_CUSTOM_GET_TOASTERID(value);
		toaster = GetTsrRoutine(toasterid);

		tattrs = palloc(sizeof(ToastAttributesData));
		tattrs->attnum = -1;
		tattrs->toasteroid = InvalidOid;
		tattrs->toastreloid = InvalidOid;
		tattrs->toasthandleroid = toasterid;
		tattrs->toaster = toaster;

		result = toaster->detoast(toast_ptr, offset, length, tattrs);
		pfree(tattrs);
	}

   return result;
}

static Datum toastapi_update (Relation rel,
												  int options,
												  Datum new_value,
												  Datum old_value,
												  int attnum)
{
	Datum result = (Datum) 0;
	struct varlena *value = (struct varlena *) DatumGetPointer(new_value);

	if(VARATT_IS_EXTERNAL_ONDISK(value))
	{
		varatt_external *old_data = (varatt_external *) DatumGetPointer(new_value);
		varatt_external *new_data = (varatt_external *) DatumGetPointer(old_value);

		if (old_data->va_toastrelid == new_data->va_toastrelid
			&& old_data->va_valueid == new_data->va_valueid)
		{
			value = (struct varlena *) detoast_attr((struct varlena *) DatumGetPointer(new_value));

			if(value)
			{
				toast_update_datum(old_value,
							   value,
								0,
								new_data->va_rawsize,
							   NULL,
								0,
								NULL,
								NULL,
								options);

				toast_delete_datum(rel, old_value, false);
			}
		}
	}
	else if(VARATT_IS_CUSTOM(value))
	{
		TsrRoutine *toaster = NULL;
		Oid	old_toasterid = InvalidOid;
		Oid	new_toasterid = InvalidOid;
		ToastAttributes tattrs;
		
		new_toasterid = VARATT_CUSTOM_GET_TOASTERID(value);

		value = (struct varlena *) DatumGetPointer(old_value);
		old_toasterid = VARATT_CUSTOM_GET_TOASTERID(value);

		if(new_toasterid != old_toasterid)
			ereport(ERROR, (errcode(ERRCODE_INVALID_CHARACTER_VALUE_FOR_CAST),
								errmsg("New value toast handler %u does not match old value handler %u",
				new_toasterid, old_toasterid)));
		
		toaster = GetTsrRoutine(new_toasterid);

		tattrs = palloc(sizeof(ToastAttributesData));
		tattrs->ntoasters = 0;
		tattrs->toasteroid = InvalidOid;
		tattrs->toastreloid = InvalidOid;
		tattrs->attnum = attnum;
		tattrs->toasthandleroid = new_toasterid;
		tattrs->toaster = toaster;

		result = toaster->update_toast(rel, new_toasterid, options, new_value, old_value, attnum, tattrs);
		toaster->deltoast(rel, old_value, false, tattrs);
		pfree(tattrs);
	}
		
   return result;
}

static Datum toastapi_copy (Relation rel,
									Datum copy_value,
									bool is_speculative,
									int attnum)
{
	Datum result = copy_value;
	struct varlena *value = (struct varlena *) DatumGetPointer(copy_value);

	if (VARATT_IS_EXTERNAL_ONDISK(value))
	{
		Datum		detoasted_newval;

		detoasted_newval = PointerGetDatum(detoast_attr(value));
		result = toast_save_datum(rel, detoasted_newval,
										  NULL, 0);
	}
	else if(VARATT_IS_CUSTOM(copy_value))
	{
		TsrRoutine *toaster = NULL;
		Oid	toasterid = InvalidOid;
		ToastAttributes tattrs;

		toasterid = VARATT_CUSTOM_GET_TOASTERID(value);
		
		toaster = GetTsrRoutine(toasterid);

		tattrs = palloc(sizeof(ToastAttributesData));
		tattrs->ntoasters = 0;
		tattrs->toasteroid = InvalidOid;
		tattrs->toastreloid = InvalidOid;
		tattrs->attnum = attnum;
		tattrs->toasthandleroid = toasterid;
		tattrs->toaster = toaster;

		if(toaster->copy_toast)
			result = toaster->copy_toast(rel, toasterid, copy_value, 0, attnum, tattrs);
		pfree(tattrs);
	}

   return result;
}

static Datum toastapi_delete (Relation rel,
										Datum del_value,
										bool is_speculative,
										int attnum)
{
	Datum result = (Datum) 0;
	struct varlena *value = (struct varlena *) DatumGetPointer(del_value);

	if(VARATT_IS_EXTERNAL_ONDISK(value))
	{
		toast_delete_datum(rel, del_value, is_speculative);
	}
	if(VARATT_IS_CUSTOM(value))
	{
		TsrRoutine *toaster = NULL;
		Oid	toasterid = InvalidOid;
		ToastAttributes tattrs;
		
		toasterid = VARATT_CUSTOM_GET_TOASTERID(value);
		
		toaster = GetTsrRoutine(toasterid);

		tattrs = palloc(sizeof(ToastAttributesData));
		tattrs->ntoasters = 0;
		tattrs->toasteroid = InvalidOid;
		tattrs->toastreloid = InvalidOid;
		tattrs->attnum = attnum;
		tattrs->toasthandleroid = toasterid;
		tattrs->toaster = toaster;

		toaster->deltoast(rel, del_value, is_speculative, tattrs);
		pfree(tattrs);
	}

   return result;
}

bool get_toast_params(Oid relid, int attnum, ToastAttributes tattrs) // int *ntoasters, Oid *toasteroid, Oid *toastrelid, Oid *handlerid)
{
	Datum d;
	char str[12];
	char *ntoasters_str;
	int len = 0;
	bool all_found_ind = true;

/*
	*ntoasters = 0;
	*toasteroid = InvalidOid;
	*toastrelid = InvalidOid;
	*handlerid = InvalidOid;
*/
	str[0] = '\0';

	d = attopts_get_toaster_opts(relid, "", attnum, ATT_NTOASTERS_NAME);
	if(d == (Datum) 0)
		all_found_ind = false;
	else
	{
		ntoasters_str = DatumGetCString(d);
		tattrs->ntoasters = atoi(ntoasters_str);
		// len = pg_ltoa(*ntoasters, str);
	}

	d = get_complex_att_opt(relid, ATT_HANDLER_NAME, str, len, attnum);
	if(d == (Datum) 0)
		all_found_ind = false;
	else
		tattrs->toasthandleroid = atoi(DatumGetCString(d));

	d = get_complex_att_opt(relid, ATT_TOASTER_NAME, str, len, attnum);
	if(d == (Datum) 0)
		all_found_ind = false;
	else
		tattrs->toasteroid = atoi(DatumGetCString(d));

	d = get_complex_att_opt(relid, ATT_TOASTREL_NAME, str, len, attnum);
	if(d == (Datum) 0)
		all_found_ind = false;
	else
		tattrs->toastreloid = atoi(DatumGetCString(d));

	return all_found_ind;
}

void _PG_init(void)
{
	create_pg_toaster();
	// create_pg_toastrel();
   toastapi_init_hook = Toastapi_init_hook;
   toastapi_toast_hook = Toastapi_toast_hook;
   toastapi_detoast_hook = Toastapi_detoast_hook;
   toastapi_size_hook = Toastapi_size_hook;
	toastapi_copy_hook = Toastapi_copy_hook;
	toastapi_update_hook = Toastapi_update_hook;
	toastapi_delete_hook = Toastapi_delete_hook;

   Toastapi_init_hook = toastapi_init;
   Toastapi_toast_hook = toastapi_toast;
   Toastapi_detoast_hook = toastapi_detoast;
   Toastapi_size_hook = toastapi_size;
	Toastapi_copy_hook = toastapi_copy;
	Toastapi_update_hook = toastapi_update;
	Toastapi_delete_hook = toastapi_delete;
}

void _PG_fini(void)
{
   Toastapi_init_hook = toastapi_init_hook;
   Toastapi_toast_hook = toastapi_toast_hook;
   Toastapi_detoast_hook = toastapi_detoast_hook;
	Toastapi_copy_hook = toastapi_copy_hook;
	Toastapi_update_hook = toastapi_update_hook;
	Toastapi_delete_hook = toastapi_delete_hook;
	Toastapi_size_hook = toastapi_size_hook;
}