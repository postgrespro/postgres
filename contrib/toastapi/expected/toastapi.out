CREATE EXTENSION toastapi;
SET search_path = pgpro_toast, public, pg_catalog;
CREATE FUNCTION dummy_toaster_handler(internal)
RETURNS internal
AS '$libdir/toastapi'
LANGUAGE C;
-- add_toaster()
SELECT add_toaster(NULL, NULL);
 add_toaster 
-------------
            
(1 row)

SELECT add_toaster(NULL, 'foo');
 add_toaster 
-------------
            
(1 row)

SELECT add_toaster('bar', NULL);
 add_toaster 
-------------
            
(1 row)

SELECT add_toaster('', '');
ERROR:  invalid name syntax
SELECT add_toaster('foo', '');
ERROR:  invalid name syntax
SELECT add_toaster('foo', 'bar');
ERROR:  function bar(internal) does not exist
SELECT add_toaster('dummy', 'foo');
ERROR:  function foo(internal) does not exist
SELECT add_toaster('dummy', 'dummy_toaster_handler') AS dummy_toaster_oid
\gset
SELECT :dummy_toaster_oid <> 0;
 ?column? 
----------
 t
(1 row)

SELECT add_toaster('dummy', 'foo');
ERROR:  function foo(internal) does not exist
SELECT :dummy_toaster_oid = add_toaster('dummy', 'dummy_toaster_handler');
 ?column? 
----------
 t
(1 row)

SELECT add_toaster('foo', 'dummy_toaster_handler') AS foo_toaster_oid
\gset
SELECT :foo_toaster_oid <> 0;
 ?column? 
----------
 t
(1 row)

SELECT :foo_toaster_oid <> :dummy_toaster_oid;
 ?column? 
----------
 t
(1 row)

SELECT :foo_toaster_oid = add_toaster('foo', 'dummy_toaster_handler');
 ?column? 
----------
 t
(1 row)

-- drop_toaster()
SELECT drop_toaster(NULL);
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('');
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('bar');
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('foo') = :foo_toaster_oid;
 ?column? 
----------
 t
(1 row)

SELECT drop_toaster('foo');
 drop_toaster 
--------------
             
(1 row)

CREATE TABLE tab (id int, jb jsonb, b_comp bytea, b_uncomp bytea STORAGE external);
-- set_toaster()
SELECT set_toaster('', '', '');
ERROR:  invalid name syntax
SELECT set_toaster('foo', 'bar', 'baz');
ERROR:  relation "bar" does not exist
SELECT set_toaster('foo', 'tab', 'baz');
ERROR:  cannot find toaster with name "foo"
SELECT set_toaster('dummy', 'tab', 'bar');
ERROR:  column "bar" of relation "tab" does not exist
SELECT set_toaster('dummy', 'tab', 'id');
ERROR:  data type integer can not be toasted
SELECT set_toaster('dummy', 'tab', 'jb');
ERROR:  toaster "dummy" does not support type jsonb
SELECT set_toaster('dummy', 'tab', 'b_uncomp');
ERROR:  toaster "dummy" supports only STORAGE EXTENDED
SELECT set_toaster('dummy', 'tab', 'b_comp') = :dummy_toaster_oid;
 ?column? 
----------
 t
(1 row)

-- get_toaster()
SELECT get_toaster('', '');
ERROR:  invalid name syntax
SELECT get_toaster('foo', 'bar');
ERROR:  relation "foo" does not exist
SELECT get_toaster('tab', 'bar');
ERROR:  column "bar" of relation "tab" does not exist
SELECT get_toaster('tab', 'id');
 get_toaster 
-------------
            
(1 row)

SELECT get_toaster('tab', 'jb');
 get_toaster 
-------------
            
(1 row)

SELECT get_toaster('tab', 'b_uncomp');
 get_toaster 
-------------
            
(1 row)

SELECT get_toaster('tab', 'b_comp') = :dummy_toaster_oid;
NOTICE:  dummy
 ?column? 
----------
 t
(1 row)

-- reset_toaster()
SELECT reset_toaster('', '');
ERROR:  invalid name syntax
SELECT reset_toaster('foo', 'bar');
ERROR:  relation "foo" does not exist
SELECT reset_toaster('tab', 'bar');
ERROR:  column "bar" of relation "tab" does not exist
SELECT reset_toaster('tab', 'id');
 reset_toaster 
---------------
             0
(1 row)

SELECT reset_toaster('tab', 'jb');
 reset_toaster 
---------------
             0
(1 row)

SELECT reset_toaster('tab', 'b_uncomp');
 reset_toaster 
---------------
             0
(1 row)

SELECT reset_toaster('tab', 'b_comp');
 reset_toaster 
---------------
             0
(1 row)

SELECT get_toaster('tab', 'b_comp');
 get_toaster 
-------------
            
(1 row)

SELECT reset_toaster('tab', 'b_comp');
 reset_toaster 
---------------
             0
(1 row)

SELECT get_toaster('tab', 'b_comp');
 get_toaster 
-------------
            
(1 row)

SELECT set_toaster('dummy', 'tab', 'b_comp') = :dummy_toaster_oid;
 ?column? 
----------
 t
(1 row)

SELECT get_toaster('tab', 'b_comp') = :dummy_toaster_oid;
NOTICE:  dummy
 ?column? 
----------
 t
(1 row)

SELECT reset_toaster('tab', 'b_comp');
 reset_toaster 
---------------
             0
(1 row)

SELECT get_toaster('tab', 'b_comp');
 get_toaster 
-------------
            
(1 row)

SELECT set_toaster('dummy', 'tab', 'b_comp') = :dummy_toaster_oid;
 ?column? 
----------
 t
(1 row)

-- drop_toaster()
SELECT drop_toaster(NULL);
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('');
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('foo');
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('dummy');
 drop_toaster 
--------------
             
(1 row)

SELECT drop_toaster('dummy');
 drop_toaster 
--------------
             
(1 row)

SELECT reset_toaster('tab', 'b_comp');
 reset_toaster 
---------------
             0
(1 row)

SELECT drop_toaster('dummy') = :dummy_toaster_oid;
 ?column? 
----------
 t
(1 row)

SELECT drop_toaster('dummy');
 drop_toaster 
--------------
             
(1 row)

DROP EXTENSION toastapi;
